### PR TITLE
Add storage for Aliyun Block Storage

### PR DESCRIPTION
**Pull request by thxCode**

### CODE REVIEW

**Comment by zionwu** [comment]
rancher/rancher#9194

File: package/abs/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,27 @@
+############################################################
+# Dockerfile to build Rancher Aliyun Block Storage Containers
+# Based on Ubuntu
+############################################################
+
+# Set the base image to Ubuntu 16.04
+FROM ubuntu:16.04
+
+# Change apt source
+# COPY abs/aliyun.list /etc/apt/sources.list.d/
```
Comment (yasker) [inline package/abs/Dockerfile]: Commented out? Then why do we need `aliyun.list`?

File: package/abs/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,27 @@
+############################################################
+# Dockerfile to build Rancher Aliyun Block Storage Containers
+# Based on Ubuntu
+############################################################
+
+# Set the base image to Ubuntu 16.04
```
Comment (yasker) [inline package/abs/Dockerfile]: It's too obvious, you don't need to comment.

File: package/abs/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+FROM ubuntu:16.04
+
+# Change apt source
+# COPY abs/aliyun.list /etc/apt/sources.list.d/
+
+# Update repo
```
Comment (yasker) [inline package/abs/Dockerfile]: Same here, and all the followings..

File: package/abs/aliyun.list [hunk 4]
```diff
@@ -0,0 +1,19 @@
+# deb cdrom:[Ubuntu 16.04 LTS _Xenial Xerus_ - Release amd64 (20160420.1)]/ xenial main restricted
```
Comment (yasker) [inline package/abs/aliyun.list]: Don't know if we still need this due to the commented out part in Dockerfile.

File: package/abs/rancher-abs [hunk 5]
```diff
@@ -0,0 +1,464 @@
+#!/bin/bash
+if [ -e "$(dirname $0)/common.sh" ]; then
+    source $(dirname $0)/common.sh
+elif [ -e "$(dirname $0)/../common/common.sh" ]; then
+    source $(dirname $0)/../common/common.sh
+fi
+
+readonly ACCESS_KEY_ID=${ECS_ACCESS_KEY_ID}
```
Comment (yasker) [inline package/abs/rancher-abs]: What's this `ECS_ACCESS_KEY_ID` and `SECRET`? I didn't see it mentioned anywhere.

File: package/abs/rancher-abs [hunk 6]
```diff
@@ -0,0 +1,464 @@
...
+
+readonly LOOP_CHECK_TIMES=30
+readonly WAIT_SLEEP_TIME_IN_SECONDS=1
+
+TRANSCATION_SEQ=""
+INSTANCE_ID=""
+DISK_ID=""
+
+
+LOG() {
```
Comment (yasker) [inline package/abs/rancher-abs]: Don't use upper case function names. You can refer to https://google.github.io/styleguide/shell.xml

File: package/abs/rancher-abs [hunk 7]
```diff
@@ -0,0 +1,464 @@
...
+        aliyun_call_result=${aliyun_call_result#'/rancher_abs/'}
+    else
+        aliyun_call_result=""
+    fi
+
+    echo "${aliyun_call_result}"
+}
+
+
+# Gen clientToken
```
Comment (yasker) [inline package/abs/rancher-abs]: Too obvious.

File: package/abs/rancher-abs [hunk 8]
```diff
@@ -0,0 +1,464 @@
...
+GEN_CLIENT_TOKEN() {
+    local time_stampe=$(date +%s%N)
+    local random_char=$(cat /dev/urandom | tr -dc 'a-zA-Z' | fold -w5 | head -n1)
+
+    echo "${random_char}_${time_stampe}"
+}
+
+# Gen diskName
```
Comment (yasker) [inline package/abs/rancher-abs]: Comments are too obvious. Let code speak for itself. Don't add comments unless necessary.

File: package/abs/rancher-abs [hunk 9]
```diff
@@ -0,0 +1,464 @@
...
+        let loop_check_times=loop_check_times+1
+    done
+
+    if [ "${cur_status}"x != "${end_status}"x ]; then
+        print_error "Failed Disk ${DISK_ID} status check, expected end status is: ${end_status}, got ${cur_status}, start on ${start_status}"
+    fi
+}
+
+
+# Find avaiable device path in os, sush as '/dev/vdb', '/dev/vdc',..., '/dev/vdz'
```
Comment (yasker) [inline package/abs/rancher-abs]: Typo. `such`, not `sush`

File: package/abs/rancher-abs [hunk 10]
```diff
@@ -0,0 +1,464 @@
...
+    size=${OPTS['size']:-"5"}
+    disk_category=${OPTS['diskCategory']:-"cloud"}
+    description=${OPTS['description']}
+    snapshot_id=${OPTS['snapshotId']}
+
+    # - Generate a client_token
+    client_token=$(GEN_CLIENT_TOKEN)
+
+    # - Append command options
+    local aliyun_options="--ZoneId ${zone_id} --DiskName ${disk_name} --Size ${size} --DiskCategory ${disk_category}"
```
Comment (yasker) [inline package/abs/rancher-abs]: Inconsistent style. For `zone_id` you separated the define and assignment. Here you define it locally. You should stick to one style. Define it when using is fine with bash.

**Review by yasker — CHANGES_REQUESTED** [review]
I will review more after these comments are addressed.

File: package/abs/Dockerfile (hunk 1 above)
Comment (thxCode) [reply package/abs/Dockerfile]: I forget to delete it. This file can help me to improve the speed when updating Ubuntu at home.

File: package/abs/rancher-abs (hunk 5 above)
Comment (thxCode) [reply package/abs/rancher-abs]: They come from [catalog](https://github.com/thxCode/rancher-catalog/blob/04d993a6141097a3f8cf8e8879876b69e185be36/infra-templates/abs/0/docker-compose.yml#L12).

File: package/abs/rancher-abs (hunk 6 above)
Comment (thxCode) [reply package/abs/rancher-abs]: OK, I get.

**Comment by yasker** [comment]
Please squash the commits to make it easier to review and merge.

In general, commits should be separated in a logical way to help reviewer and a later developer to understand. In this case, one commit should be enough.

**Review by yasker — APPROVED** [review]

**Comment by yasker** [comment]
Hi @cjellick 

Can you review this one?

Thanks!

**Comment by cjellick** [comment]
LGTM. I just verified that the code can be merged without affecting api/existing functionality.

**Pull request merged**